File: apps/android/src/main/kotlin/space/bitos/app/ui/designsystem/AppSkeleton.kt
package space.bitos.app.ui.designsystem

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import space.bitos.app.ui.components.HexShape
import space.bitos.app.ui.theme.BitOSRadius
import space.bitos.app.ui.theme.bitOSColors

/**
 * Skeleton block (§2.5 "Skeletons for every list/grid"): surface→overlay
 * shimmer sweep driven by [rememberShimmerPhase]. Stack freely — the
 * convenience rows below cover the common list/grid cases.
 */
@Composable
fun AppSkeleton(
    modifier: Modifier = Modifier,
    cornerRadius: Dp = BitOSRadius.sm,
    height: Dp? = null,
) {
    val colors = bitOSColors()
    val phase = rememberShimmerPhase()
    val sweep = 1600f
    val brush = Brush.linearGradient(
        colors = listOf(colors.surfaceElevated, colors.surfaceOverlay, colors.surfaceElevated),
        start = Offset(phase * sweep - sweep / 2, 0f),
        end = Offset(phase * sweep - sweep / 2 + sweep / 3, 320f),
    )
    Box(
        modifier = modifier
            .then(if (height != null) Modifier.height(height) else Modifier)
            .clip(RoundedCornerShape(cornerRadius))
            .background(brush),
    )
}

/** Two-line list-row skeleton with a circular leading glyph. */
@Composable
fun AppSkeletonRow(modifier: Modifier = Modifier) {
    Row(
        modifier = modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 12.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        AppSkeleton(modifier = Modifier.size(40.dp), cornerRadius = 999.dp)
        Column(verticalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.weight(1f)) {
            AppSkeleton(modifier = Modifier.fillMaxWidth(0.42f), height = 12.dp)
            AppSkeleton(modifier = Modifier.fillMaxWidth(0.68f), height = 10.dp)
        }
    }
}

/** Grid tile skeleton for explore/mosaic loading states. */
@Composable
fun AppSkeletonTile(modifier: Modifier = Modifier) {
    AppSkeleton(
        modifier = modifier.fillMaxWidth().height(220.dp),
        cornerRadius = BitOSRadius.md,
    )
}

/**
 * Matched empty state (§2.5): hex icon plate + title + one-line message +
 * a single primary CTA — the "explain why and offer one primary next
 * action" rule from ux-ui-flows §3.
 */
@Composable
fun AppEmptyState(
    title: String,
    message: String? = null,
    modifier: Modifier = Modifier,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null,
    icon: @Composable () -> Unit = { DefaultEmptyIcon() },
) {
    Column(
        modifier = modifier.fillMaxWidth().padding(horizontal = 32.dp, vertical = 40.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        icon()
        Text(
            text = title,
            style = MaterialTheme.typography.headlineSmall,
            textAlign = TextAlign.Center,
        )
        if (message != null) {
            Text(
                text = message,
                style = MaterialTheme.typography.bodyMedium,
                color = bitOSColors().textSecondary,
                textAlign = TextAlign.Center,
            )
        }
        if (actionLabel != null && onAction != null) {
            Button(onClick = onAction, shape = CircleShape) {
                Text(actionLabel)
            }
        }
    }
}

@Composable
private fun DefaultEmptyIcon() {
    val colors = bitOSColors()
    Box(
        modifier = Modifier
            .size(56.dp)
            .clip(HexShape())
            .background(colors.primary.copy(alpha = 0.16f)),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = "⬡",
            style = MaterialTheme.typography.headlineMedium,
            color = colors.primaryDark,
        )
    }
}

File: apps/ios/BitOS/DesignSystem/AppSkeleton.swift
import SwiftUI

/**
 * Skeleton block (§2.5 "Skeletons for every list/grid"): elevated→overlay
 * shimmer sweep. The animation honors `accessibilityReduceMotion` by
 * rendering the static base instead of sweeping.
 */
struct AppSkeleton: View {
    var height: CGFloat? = nil
    var cornerRadius: CGFloat = BitOSTheme.Radius.sm
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var phase: CGFloat = -1

    var body: some View {
        let shape = RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
        shape
            .fill(BitOSTheme.surfaceElevated)
            .frame(height: height)
            .overlay(
                // Sweep highlight: clipped to the block, translating left→right.
                GeometryReader { geo in
                    LinearGradient(
                        colors: [
                            BitOSTheme.surfaceElevated,
                            BitOSTheme.surfaceOverlay,
                            BitOSTheme.surfaceElevated,
                        ],
                        startPoint: .leading,
                        endPoint: .trailing
                    )
                    .frame(width: geo.size.width * 1.6)
                    .offset(x: phase * geo.size.width * 1.6)
                }
                .clipShape(shape)
            )
            .onAppear {
                guard !reduceMotion else { return }
                withAnimation(
                    .linear(duration: 1.4)
                    .repeatForever(autoreverses: false)
                ) {
                    phase = 0.6
                }
            }
            .accessibilityHidden(true)
    }
}

/// Two-line list-row skeleton with a hex leading glyph.
struct AppSkeletonRow: View {
    var body: some View {
        HStack(spacing: BitOSTheme.Spacing.md) {
            AppSkeleton(height: 40, cornerRadius: BitOSTheme.Radius.md)
                .frame(width: 40)
            VStack(alignment: .leading, spacing: BitOSTheme.Spacing.sm) {
                AppSkeleton(height: 12)
                    .frame(width: 140)
                AppSkeleton(height: 10)
                    .frame(width: 210)
            }
        }
        .padding(.horizontal, BitOSTheme.Spacing.base)
        .padding(.vertical, BitOSTheme.Spacing.md)
    }
}

/**
 * Matched empty state (§2.5): hex icon plate + title + message + one
 * primary CTA — the "explain why and offer one primary next action"
 * rule from ux-ui-flows §3.
 */
struct AppEmptyState: View {
    let title: String
    var message: String? = nil
    var actionTitle: String? = nil
    var onAction: (() -> Void)? = nil
    var icon: () -> AnyView = {
        AnyView(
            Text("⬡")
                .font(BitOSType.headlineMedium)
                .foregroundStyle(BitOSTheme.textLink)
        )
    }

    var body: some View {
        VStack(spacing: BitOSTheme.Spacing.md) {
            ZStack {
                HexShape()
                    .fill(BitOSTheme.accent.opacity(0.16))
                    .frame(width: 56, height: 56)
                icon()
            }
            .frame(width: 56, height: 56)
            Text(title)
                .font(BitOSType.headlineSmall)
                .multilineTextAlignment(.center)
            if let message {
                Text(message)
                    .font(BitOSType.bodyMedium)
                    .foregroundStyle(BitOSTheme.textSecondary)
                    .multilineTextAlignment(.center)
            }
            if let actionTitle, let onAction {
                AppButton(title: actionTitle, action: onAction)
                    .padding(.top, BitOSTheme.Spacing.xs)
            }
        }
        .padding(.horizontal, BitOSTheme.Spacing.xxxl)
        .padding(.vertical, BitOSTheme.Spacing.xxxl)
        .frame(maxWidth: .infinity)
    }
}

#Preview("Skeletons + empty state") {
    ScrollView {
        VStack(spacing: BitOSTheme.Spacing.base) {
            AppSkeletonRow()
            AppSkeletonRow()
            AppSkeleton(height: 180, cornerRadius: BitOSTheme.Radius.md)
            AppEmptyState(
                title: "Nothing saved yet",
                message: "Bookmarks live on your relays (kind 10000 set) — private to you, portable across apps.",
                actionTitle: "Browse the feed",
                onAction: {}
            )
        }
        .padding(BitOSTheme.Spacing.screen)
    }
    .background(BitOSTheme.background)
    .environment(\.colorScheme, .dark)
}
